### PR TITLE
Simplifying the code declaring the constants bound to primitive projections

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** cleanup

Added: overlay [rewriter PR#3](https://github.com/mit-plv/rewriter/pull/3) (not needed anymore)

We factorize code from `declareInd.ml` and `inductiveops.ml` with code essentially already existing in `record.ml`.

This moves the declarations of constants bound to projections from `DeclareInd.declare_mutual_inductive_with_eliminations` to `Record.declare_projections` and the gain is the factorization with the code used for non-primitive projections. 

Maybe some further simplifications are possible?

I made one semantic change: in this PR, projecting constants are declared for local definitions of primitive records, the same way as they are declared for non-primitive records. As a consequence, in `Set Primitive Projections. Record T := {a:=0;b:a=a}. Print b.`, the type of `b` is now `forall t, a t = a t`. [Added: This change is finally removed, see last messages of the discussion.]

### CODE REVIEW

File: vernac/declareInd.ml [hunk 1]
```diff
@@ -145,7 +113,10 @@ let declare_mind mie =
   let (sp,kn as oname) = Lib.add_leaf id (inInductive { ind_names = names }) in
   if is_unsafe_typing_flags() then feedback_axiom ();
   let mind = Global.mind_of_delta_kn kn in
-  let isprim = declare_projections mie.mind_entry_universes mind in
+  let isprim =
+    let open Declarations in
+    let record_info = (Environ.lookup_mind mind (Global.env())).mind_record in
+    match record_info with PrimRecord _ -> true | FakeRecord | NotRecord -> false in
```
Comment (SkySkimmer) [inline vernac/declareInd.ml]: Maybe use `Inductive.is_primitive_record`?

File: pretyping/inductiveops.mli [hunk 2]
```diff
@@ -211,10 +211,6 @@ val make_case_or_project :
 val make_default_case_info : env -> case_style -> inductive -> case_info
 i*)
 
-val compute_projections : Environ.env -> inductive -> (constr * types) array
```
Comment (ppedrot) [inline pretyping/inductiveops.mli]: I am not sure this is a good idea to remove this function from the API. It was somewhat nice to have for plugins, and it is bound to be reimplemented (with its share of bugs) next time we need to compute projections for something else in the codebase.

File: vernac/record.ml [hunk 3]
```diff
@@ -322,46 +322,39 @@ let declare_projections indsp ctx ?(kind=Decls.StructureComponent) binder_name f
...
-                with Type_errors.TypeError (ctx,te) ->
-                  raise (NotDefinable (BadTypedProj (fid,ctx,te)))
+                if is_local_assum decl && primitive then
+                  let _ = DeclareInd.declare_primitive_projection p kn in
+                  let kn' = Projection.Repr.constant p in
+                  kn', mkProj (Projection.make p false,mkRel 1)
+                else
+                  let proj_args = (*Rel 1 refers to "x"*) paramargs@[mkRel 1] in
+                  kn, applist (mkConstU (kn,u),proj_args)
+              with Type_errors.TypeError (ctx,te) when not primitive ->
```
Comment (SkySkimmer) [inline vernac/record.ml]: Shouldn't we `try` only around the `declare_constant` rather than the whole block?
ie `let kn = try declare_constant ... with TypeError ... -> raise (NotDefinable ...) in`

File: pretyping/inductiveops.mli (hunk 2 above)
Comment (herbelin) [reply pretyping/inductiveops.mli]: This function is actually already contained in `Record.declare_projections`. Would you be ok that it is `record.ml` which exports a `compute_projections` function?

Alternatively, the code in `Record.declare_projections` could be split into two parts, a part which computes projections in `inductiveops.ml` and a part which declares the projections, remaining in `record.ml` (or moved to `declareInd.ml` actually), for both primitive and non-primitive projections?

File: vernac/declareInd.ml (hunk 1 above)
Comment (herbelin) [reply vernac/declareInd.ml]: Good idea, will do.

File: vernac/record.ml (hunk 3 above)
Comment (herbelin) [reply vernac/record.ml]: That's reasonable. I guess noone would object and I will do.

**Comment by SkySkimmer** [comment]
Pushing some changes as separate commits, feel free to revert if you don't like them.

**Comment by herbelin** [comment]
> Pushing some changes as separate commits, feel free to revert if you don't like them.

I like the first one. I like the second one too but I'm not totally sure that the `kn'` and the `kn` are the same. It was my assumption that they are the same, but when I tried to `assert (kn=kn')` it failed (and I stopped investigating more).

About the third one, there is an underlying question first: why not to use a `Proj` also for the projections corresponding to a let-in? There is no theoretical need for it of course, but that would allow to have a compact representation also for the (trivial) projection associated to a let-in.

**Comment by SkySkimmer** [comment]
>About the third one, there is an underlying question first: why not to use a Proj also for the projections corresponding to a let-in? There is no theoretical need for it of course, but that would allow to have a compact representation also for the (trivial) projection associated to a let-in.

Currently Proj can't be used for the letin projections. If we change that in the future we will have to change declare_projections regardless of whether we keep my commit or if we drop it, so it's no argument for dropping it.

**Comment by SkySkimmer** [comment]
>when I tried to assert (kn=kn') it failed (and I stopped investigating more).

Did you use literally `=` or `Constant.equal`?

**Comment by herbelin** [comment]
> Currently Proj can't be used for the letin projections. If we change that in the future we will have to change declare_projections regardless of whether we keep my commit or if we drop it, so it's no argument for dropping it.

Sure! I just wanted to open the discussion to know if there were some objections around to eventually associate a `Proj` node also to let-in.

**Comment by herbelin** [comment]
I used `=`.

File: pretyping/inductiveops.mli (hunk 2 above)
Comment (ppedrot) [reply pretyping/inductiveops.mli]: I think I prefer the second option. We should not put generic code for term manipulation in vernac.

Comment (herbelin) [reply pretyping/inductiveops.mli]: I agree. Then, there remains the question in `inductiveops.ml` of factorizing or not the code for primitive and non-primitive projections. There is a slight difference: the type of projection `i` refers to the _predictible_ internal name `Proj(... mind ... i ...)` in the primitive case while in the non-primitive case, it refers to a name `Const(cst)` where `cst` is only (in principle) known lately from `declare_constant`. (We can cheat though, and assume that `cst` will precisely follow the rule of being of the `Projection.Repr.constant` form, the same way the code that currently uses `Projection.Repr.constant` already cheats by anticipating that the non-compact defined projection associated to a primitive projection and obtained by `declare_constant` will have this name.)

PS: I would not like to diverge too much, but afaicj the `Proj` nodes have still no concrete syntax associated to them, so, we are actually defining types of projections which refer to terms parsable only up to delta-conversion.

Comment (herbelin) [reply pretyping/inductiveops.mli]: @ppedrot, I see two possibilities:
- if we want to keep the type of `Inductiveops.compute_projections` exactly as it is and making it work for non-primitive projections. However, I'm not too enthousiastic about preserving exactly `Inductiveops.compute_projections` for the reasons given above that it is morally interleaved with the declaration of the `Const` names in the environment.
- An alternative, which concentrates on the technical and error-prone computation of the body of the projection of index `proj_arg` is to have a function `compute_projection_body : env -> subst:constr option list -> inductive -> ~proj_arg:int -> (constr * types) option`  where `None` in the two occurrences of `option` is to tell which previous projections (in `subst`) were not definable and to tell that the demanded projection cannot be defined, respectively (I use here labels for explanation purpose, not to be kept in code).

In both cases, `FakeRecord` would now carry a default name for inhabitants of the record, as it is done for `PrimRecord`.

Does it fit your view? Or else, what was your own idea?

Comment (ppedrot) [reply pretyping/inductiveops.mli]: Please don't add information to `FakeRecord`, I have been trying to get rid of it for a long time and was blocked due to interaction with the semantics of functors. This stuff shouldn't be in the kernel at all, this is really due to libobject quirks. Currently IIRC it's only relied on by extraction.

I don't think the Inductiveops function should work over non-primitive records. It has been designed precisely for this use case, and is a low-level utility function that is hard to get right, just as you wouldn't reimplement say DeBruijn substitution everywhere. Why can't you split depending on the primitive status in the upper layers?

Comment (herbelin) [reply pretyping/inductiveops.mli]: The original reason I came to this PR is that I'm investigation how to attach manual implicit arguments to `declare_constant` rather than having `declare_constant` first declaring the automatic implicit arguments, then `declare_manual_implicit` overriding this computation with a new implicit object including the manual implicit arguments. So, my motivations were: more locality, less side-effects, more intuitive.

Now, along this exploration, I stumbled on `declare_constant` in `Record.declare_projections` which is not done at the same time in the primitive and in the non-primitive case. If found this a difficult invariant to follow, so I tried to move the `declare_constant` from `declareInd.ml` to `Record.declare_projections`. In the way, I realized that the code for non-primitive projections in `declare_projections` was the same as the code of primitive projections in `inductiveops.ml`. So, I factorized it and the resulting code was indeed simpler.

So, to answer precisely this question:

> Why can't you split depending on the primitive status in the upper layers?

There is already a split depending on the primitive status, but the fact is that the split is not on the technical part involving de Bruijn indices. The split is only on whether we use a `Proj` node or a `Case` node in the body and a `Proj` node or `Const` node in the dependencies.

----

Now, I guess that the underlying question is where we are going. If you consider `FakeRecord` to be temporary, I guess that this means that you consider the code in `Record.declare_projections` to be eventually removed and all declarations introduced by the keyword `Record` based on primitive projections, right? Then, what about `Recordops.register_primitive_projection`? I guess that the `Const` bound to `Proj` nodes are also deemed to disappear, right? And that we shall have a proper user syntax for the `Proj` node instead, right? And a proper `Proj` node in `GlobRef.t` so that we can e.g. attach a scope to its argument? And that we shall eta-expand unapplied projections at parsing time?

In this case, if the point is to keep `Inductiveops.compute_projections` for future use, that's ok to me.

File: vernac/record.ml (hunk 3 above)
Comment (ejgallego) [reply vernac/record.ml]: Even better would be to remove the exception and use `result` , but that's orthogonal to this PR.

File: pretyping/inductiveops.mli (hunk 2 above)
Comment (ppedrot) [reply pretyping/inductiveops.mli]: I agree with your plan, except maybe for the η-expansion of projections at parsing time. I'd lean instead towards prohibiting the use of non-fully applied projections. In languages like ML this is even enforced syntactically, and in Coq we're paying a huge cost of trying to strike a middle ground on this topic.

Comment (herbelin) [reply pretyping/inductiveops.mli]: Do we renounce to any form of compatibility with existing code using "positive" records, such as taking benefit of primitive projections for `prod` or `sigT`?

**Comment by herbelin** [comment]
@skyskimmer: I guess it is ok if I squash your commits into a single main commit?

**Comment by SkySkimmer** [comment]
Sure.

File: pretyping/inductiveops.mli (hunk 2 above)
Comment (ppedrot) [reply pretyping/inductiveops.mli]: I was under the impression that the conclusion of the current mess is that compatibility is but a dream...

Comment (herbelin) [reply pretyping/inductiveops.mli]: The original motivation for primitive projections was efficiency. In practice, it introduced two features. On one side, it brought negative record types (i.e. sugar for `match` and eta-conversion). On the other side, it brought compact representation of projections (i.e. dropping the parameters in the internal representation). I think that the two concepts should be kept separated.

That `match t with (x,y) => term end` in primitive positive records compatibility mode was sugar for `let x := fst t in let y := snd t in term` and this has been a source of problems. This introduced confusion in printing and broke reduction. So, let's renounce to simulating positive record types from negative record types and let's instead preserve `match` in positive records as it is now.

Let us then consider different levels.

The mininum servicing is just to adopt a compact representation for the constants associated to `match`-based projections in positive records. For that we need a constructor say `CompactProj(cst,t)`, thought as `App (Const cst, inferred_parameters @[t])`, and we instrument `mkApp` so that it uses `CompactProj` when applied to such constant. In all other respects, this behaves as current positive records. The bet is that all what has to be done for such a `CompactProj` is already mostly done in the code for the `Proj` node, so that `CompactProj(cst,t)` could instead be encoded, say, as `Proj(Projection.make(...,true),t)` (where we recycle `true` in the proj node to mean  projection from a "positive" type rather than "unfolded"). I believe we would mostly have compatibility up to the exception that the record parameters are not any more part of the internal syntax and thus things like `pattern t at n` would work differently. More generally, printing would be a bit different: the parameters are inferred and the user has no more control on them.

[I suspect that this approach may eventually generalize to other compact representations of application nodes in cases where some arguments are canonically inferable.]

The next level is to support eta-conversion in positive record types. The claim (to be checked formally) is that conversion modulo match-eta-conversion in positive record types is equivalent to conversion modulo proj-eta-conversion in negative record types. Thus in the conversion, we stop treating a positive record type as positive but treat it as if it were negative. This means blocking the reduction of `Proj` and expanding `match t with (x,y) => u end` into `u[x:=fst t][v:=snd t]`, as well as applying the same eta-conversion for (finite) as we already do for negative record types. At this level, I don't think we would lose a lot of compatibility, except in some cases where a more powerful conversion may matter.

A further level would be to avoid losing the compact representation of a positive projection in reduction tactics by blocking the expansion of projections in positive record types to a `match`, though still allowing its iota-contraction when applied to a constructor. This may be strange in the sense that both `fst z` and `match z with (x,y) => x end` would be normal forms and none of `cbv`, `cbn`, etc would move one to the other. We could add however a `cbv+` which prefers expanding `fst` into `match` and a `cbv-` which prefers expanding `match` into `fst` and `snd`. At this level, compatibility is virtually lost whenever we try to `unfold` a positive projection applied to a neutral term, or more generally, whenever we ask for a delta reduction of projection which cannot be followed by a iota.

If we are interesting in this last level, we are actually not far from a convergence between (finite) negative or positive record types. For instance, we could go as far as saying that `match` is also available in negative records. This would not be a big step since something behaving like `match` already exists in negative record type under the form of `let x:=fst z in let y:=snd z in t`. The only difference is that we would provide a primitive representation for it and block its reduction when `z` is not a pair. At the reduction level, `cbv`, `cbn` and co would be different but `cbv-` would implement reduction as if `match` were indeed pure sugar. No change in the way conversion is implemented though. 

In this last level, we would not need any more the `bool` flag from above to differentiate between positive and negative types. They would both behave the same in conversion and the only difference between a (finite) negative record type and a (finite) positive record type would be at the level of the reduction: `cbv+` would implement reduction as if in a positive type; `cbv-` would implement reduction as if in a negative type.

**Comment by herbelin** [comment]
Let me reformulate and extend my last comment in a synthetic way.

1. Claim: Negative record types and compact representations of applied projections are two distinct issues.

2. Observation: Positive record types with eta for `match` are isomorphic to negative record types with eta for projections. In one way, `match z with (x,y) => t end` is expanded into `t[x:=fst z][y:=snd z]` and in the other way, `fst z` and `snd z` are expanded to `match z with (x,y) => x end` and `match z with (x,y) => y end` respectively.

3. Corollary 1: There is no difference between positive record types with eta and negative record types with eta from the point of view of conversion.

4. Corollary 2: Deciding conversion in finite posititive record types with eta comes for free by proceeding the same way as for finite negative record types with eta.
    - Expand `match z with (x,y) => t end` in `t[x:=fst z][y:=snd z]`
    - Block projections applied to a neutral term
    - Turn conversion of `v` against `(t,u)` into conversion of `fst v` against `t`, and of `snd v` against `u`

5. Claim: It is no harm to accept an explicit `match z with (x,y) => t end` implicitly expanded to `let x:=fst z in let y:=snd z in t` in negative record types, symmetrically to the current ability to define explicit projections `fst` and `snd` in positive record types that expand to `match z with (x,y) => x end` and `match z with (x,y) => y end`. For `match` in negative types, we can use the `Case` constructor (rather than the current syntactic expansion in terms of `Proj(true)`). For `fst` and `snd` in positive record types, we can use the `Proj` constructor (rather than `Const` as currently).

6. A clear picture:
    - The `Case` and `Proj` constructors of `Constr.t` are then only about the representation of `match` and projections. They do not say if the type in which they live is negative or positive. They are parsed and printed as `match` and as `t.(proj)`.
    - No intricate features governed by compatibility options: no Boolean in `Proj`, no `make_case_or_project` elaboration of `match` into projections,  use of `Proj` for everyone in `Record.declare_projections`, no `register_primitive_projection` or `Projection.Repr.constant` to connect a `Proj` to a `Const`, ...
    - Conversion with eta works the same for the positive and negative finite cases.
    - It remains how to deal with reduction.

7. Design choices in the reduction:
    - Choice 1: The difference between positive and negative record types is about whether `match` expand to projections or projections expand to `match`. Thus, `FakeRecord` disappears and we add a Boolean in `PrimRecord` telling if the type is positive or negative with impact on how it behaves in reduction.
    - Choice 2: We renounce to make a difference between positive and negative record types. Both have `match` and projections blocked in reduction when applied to neutral terms. We loose the existence of a unique normal form (even if all normal forms are convertible). `FakeRecord` disappears and `PrimRecord` remains as it is.
    - Choice 3: Idem as choice 2 but we provide new reduction rules, `iota-` which expands `match` into projections and `iota+` which expands projections into `match`. They are exclusive the one of the other.

8. Conjecture: In the coinductive case, positive record types with bridled return clause (#10764) for `match`, without eta for `match` but with commutative cuts and presumably some `match` contraction rule are isomorphic to coinductive negative record types without eta for projections.

9. Corollary: The same discussion as above would apply to the coinductive case.

10. Open question: What happens in the inductive case? It seems that eta for projections can be added to the conversion in a decidable way. However, eta for `fix` shall not be decidable. What equivalence can we draw between the positive and negative cases?

**Comment by ppedrot** [comment]
@herbelin can you rebase? I think this PR has converged, and I'll merge it afterwards.

**Comment by herbelin** [comment]
Rebased.

PS: Do you know if it is expected to get "clang: warning: argument unused during compilation: '-fexcess-precision=standard'" when compiling under MacOS?

**Comment by ppedrot** [comment]
Is the CI failure related?

**Comment by herbelin** [comment]
> Is the CI failure related?

Yes, it is related to the semantic change mentioned in the initial message:
```
Set Primitive Projections. 
Record T := {a:=0;b:a=a}.
Print b.
```
The type of `b` is not anymore `forall t, 0 = 0` but `forall t, a t = a t`, as it is without primitive projections.

I believe this change is wanted. So, in the absence of objections, I'm going to submit a patch upstream. (Incidentally, the CI is really impressive!)

**Comment by JasonGross** [comment]
Given that this PR needed no changes to the test-suite, but did need changes to the CI, I think it might be nice to add a small case to the test suite to cover this change in behavior (I think ideally the test-suite should have 100% feature coverage, and it seems like it's particularly easy to take a step towards that here).

**Comment by herbelin** [comment]
> I think it might be nice to add a small case to the test suite to cover this change in behavior

You're right.

Then, preparing a test-suite file led me to another place of the code where primitive records unfolds local definitions in the type of other fields, namely in `Indtypes.compute_projections`. A bigger change would be needed, requiring that the array of primitive projections includes the projections corresponding to local definitions (so that, say `Proj((kn,i),t)` can refer to `Proj((kn,j),t)` in its type for the j-th component being a local definition). I'm ready to do this change but this is a more intrusive one about which the opinion of (at least) @mattam82 and @ppedrot would be required.

Concrete example:
```
Set Primitive Projections. 
Record T := {a:=0; b:a=0}.                                                                                                                    
Check b.
(* Formerly: "T -> 0 = 0" *)
(* Currently: "forall t:T, a t = 0" which is computed in `record.ml` for the Const *)
Check fun t => b t. 
(* Currently: "T -> 0 = 0" which is computed in `indtypes.ml` for the Proj *)
```

In short, the current PR introduces an inconsistency between the `Const` and `Proj` versions of a primitive projection, with a design decision to take on how to resolve the inconsistency.

**Comment by herbelin** [comment]
@ppedrot (and @mattam82 if you have some time available, and everyone else):

My proposal to prevent expansion of let-in primitive projections is to canonically attach to any 
let-in field `x:=c` in a primitive record type `modpath.R` a constant of name `modpath.x` with value `c`. This means that declaring an inductive type would declare more than only an "mutual_inductive_body": it shall also declare a list of constants. Would you be ready with that? 

The alternative would be to use a `Proj` constructor to encode the let-in projections, e.g. by extending it into:
```
type t = 
    { proj_ind : inductive;         
      proj_npars : int;             
      proj_is_local_def  : bool;
      proj_arg : int;
       (* if proj_is_local_def = false, proj_arg is the index among assumption *)
       (* if proj_is_local_def = true, proj_arg is the index among definitions *)
      proj_name : Label.t; }
```

Any preference?

----

In the longer term, I may propose the following:
- have a constructor `Proj(p)` in `GlobRef.t` rather than a constructor `Proj(p,c)` in `constr`
- add a constructor `AppCompact` (or whatever name) in `constr` and replace `Proj (p,c)` in `constr` by `AppCompact (Proj(p),c)`

This would solve the "eta-expansion" problem with projections: `p` alone is `Proj(p)`, applied `p c` (or `c.(p)`) is `AppCompact (Proj(p),c)`.

In the longer longer term, the constructor `AppCompact` could be generalized to other forms of application for which skipping implicit arguments is useful.

**Comment by SkySkimmer** [comment]
Not sure we want to stick more stuff in the kernel tbh

**Comment by ppedrot** [comment]
I really don't like the idea of having compact applications in the term node that do not correspond to a typing rule. This is asking for trouble, the kernel-facing syntax should be as close as possible to both what we actually enforce and what the user sees.

**Comment by ppedrot** [comment]
Also, we have let bindings, I don't understand why we don't use them anymore. I think that in @herbelin's example given above, we should have instead
```
b : forall t, let a := 0 in a = a
```
This is more in line with the view of a record as the adjoint of a context formation.

**Comment by herbelin** [comment]
@SkySkimmer, @ppedrot: My proposal did not intend at all to make the kernel more complicated or to not reflect a typing rule.

Nevertheless, are you both saying that our priority should be to first reduce the number of features in the kernel?

**Comment by ppedrot** [comment]
> Nevertheless, are you both saying that our priority should be to first reduce the number of features in the kernel?

Maybe not reducing, but at least making the dubious ones more trustworthy (I am looking at you, modules). This also implies going against a well-known tendency in the Coq history, namely *not* adding kernel-level support for userland features. Famous examples that are still a PITA include the kername aliasing system and the unfold flag for primitive projections. As the concept implies, a kernel needs to be self-contained.

**Comment by herbelin** [comment]
> I am looking at you, modules

Indeed, modules add comparatively a lot of code. What are the plans on this side?

> the unfold flag for primitive projections

This is precisely one thing I want to address, as well as the confusions between a non-applied projection-as-`Const` and an applied one-as-`Proj` (with the hope of a _full_ compatibility so that we can finally take benefit of primitive projections in the existing libraries!).

**Comment by herbelin** [comment]
@skyskimmer, @ppedrot:

More generally, my tentative plan is the following:
- supporting efficient representation of projections also for "positive" records, using the code already existing for "negative" records;
- supporting eta for finite positive records;
- since the only essential difference between positive and negative records is reduction towards projection-based vs reduction towards match-based normal forms, the kernel difference between positive and negative records should reduce to a single boolean in the `PrimRecord` field (and removing `FakeRecord`), as outlined [above](https://github.com/coq/coq/pull/11135#issuecomment-557778473);
- since unapplied projection `p` has a kernel representation in "positive" records, that would be simpler to have it also for negative records (hence my proposal to have a `ProjRef` constructor in `GlobRef.t`); of course, such `ProjRef` will have to be thought as a compact representation of its eta-expansion in exactly the same way as `Construct` can be seen as a compact kernel representation of what is morally its eta-expansion;
- since local definitions in positive records are seen as defined fields, I would lean to do the same for negative records (and otherwise changing what is done for positive records along the lines proposed above by @ppedrot);
- removing the useless unfold flag for primitive projections;
- no particular opinion on what the syntax `t.(p)` should be bound to (in particular, with the above, `p t` and `t.(p)` are represented the same in the kernel, compactly when `p` is a projection; non compactly when `p` is an arbitrary constant);
- incidentally, I'd eventually like to also support eta for unit and empty types (by keeping a typing environment in `Reduction.ccnv` and recomputing the type of the conversion problem whenever we have a `(x|stack) = (x',stack')` conversion problem with distinct `x` and `x'`.
 
Are you ok with this plan?

**Comment by JasonGross** [comment]
I am a fan of this plan, @herbelin.  Especially the bit about judgmental eta for unit. (Perhaps I'll finally get to move forward on the category theory I've been blocking on waiting for unit with eta + anonymous record types (or at least ones whose conversion is checked based on the types of the fields) that @ppedrot mentioned implementing).

**Comment by herbelin** [comment]
@jasongross: Thanks for the support!

@skyskimmer, @ppedrot: How to make progress at least already on this PR, i.e. on the question of local definitions in records?

For positive records, projections, whether they correspond to assumptions or local definitions, are bound to constants in the kernel (`Const` nodes).

For "compactly-represented" projections (and I use the words "compactly-represented" to express that this is the compact representation which matters here and not the negative-style of so-called primitive projections which matters), projections to assumptions are represented with a `Proj` node when applied and with a `Const` node when not applied, while projections to local definitions are ignored and unfolded where they occur.

I like very much the idea that the kernel representation of objects reflect the "geometry" of the underlying logic, and, thus that `Proj` nodes are primitive "destructors" of negative types dual to the primitive "constructors" (`Construct` nodes) of positive types.

I'm not however completely convinced by the idea of repeating the local definitions of a record in the type of the projections which depend on them. Of course, this is debattable (and I'd be happy to hear more feedback around), but I feel that what is done for positive records is nice and that such repetition would defeat the purpose of local definitions in records.

If we accept the idea that the model used for positive records is good (and I will accept it in the absence of other feedback so that I can continue making progress on the question), the next question is how to represent these local-definition fields.

Here are the different solutions I could imagine:
1. change the meaning of the `proj_arg` field of the `Proj` node so that it counts the number of all fields rather than the number of assumption-fields
   - pros: easy
   - cons: this would introduce a penalty for the reduction machine which needs to know quickly which argument of the corresponding constructor has to be projected
2. add a new field, say `proj_decl : int option` to the `Proj` node which tells, when `Some i`, that the projection is a local definition, and `i` is its index among all fields (to be found in the `mind_spec`)
3. change `proj_arg : int` to `proj_arg : bool * int` where `bool * int` gives the extra information which `proj_decl` gives in the previous item
4. change the representation of `Construct` nodes for negative types so that it does not contain the arguments but the arguments + local definitions
   - cons: this makes the terms "bigger" and this seems again to defeat the purpose of local definitions
5. do something a bit more "daring"/"impure" which is to use a `Const` node instead:
   - an advantage is that a local definition field does not interact with a constructor like `Proj` does; so this seems somehow cleaner from the point of view of reduction: a `Const` simply unfolds, independently of what is in the stack
   - the drawback is that such `Const`, on the contrary of other constants, would have to be typed _mutually_ with the definition of the record type (since it would occur explicitly in the type of the field of the corresponding `mutual_inductive_body`)
      - we can then accept this kind of circularity that we don't yet have in the kernel (**)
      - or we can do like we already do for the recursivity in inductive types and "tie" the recursivity only at use time (as we do in `Inductive.constructor_instantiate`) (*)
6. add a new node `ProjDef` (or maybe `CoConst` to mean that it is a constant unfolding to a destructor), but that seems a bit overkill: local-definition fields are not an important concept enough to justify having a node just for them

(*) Incidentally, do we really need this canonical `Inductive.constructor_instantiate` each time we want to type a constructor? Why not to hard-wire the recursivity of the `Ind` field directly in the `mutual_inductive_body` rather than repeating the same instantiation every time?

(**) If we eventually get mutual global fixpoints, which users are hoping, we would naturally need a kind of declarations which defines several mutual objects at the same time.

TL;LR

My thesis is that the way definitions-fields are treated in positive records is good and worth to be supported in the kernel. How to do this support in the less intrusive way? I would lean towards 2, 3 or 5.

In any case, to make _progress_ on primitive projections (and, to my knowledge, _everyone_ is waiting for such progress, hopefully with a _compatibility_), we need to go towards the _convergence_ of positive and negative records, and in particular to resolve the question of unifying the treatment of definition-fields between positive and negative records.

**Comment by herbelin** [comment]
I added primitive projections at the menu of the next Coq call. If I correctly interpret the different views, the current open question is: Is it legitimate taking benefit of having a primitive (kernel) treatment of projections in positive record types even though projections in positive record types are not primitive from the logical point of view?

**Comment by ppedrot** [comment]
This PR has derailed. What was originally a small code factorization has now turned into a long-term project. Is the code change introduced by this PR still relevant? If not, I am closing it.

**Comment by herbelin** [comment]
@ppedrot:

Let me remind my point of view:
- The factorization between positive and negative projections made in this PR shows a difference of behavior of both kinds of records wrt local definitions.
- I could renounce to unify these behavior and complete the PR.
- I did not do it because addressing the question of non-uniformity of local definitions between the two representations appears to me as a useful _incent_ to not postpone a new time the completion of the design of primitive projections.
- I thought that completing the design of primitive projections was a common wish of most developers; am I correct here?
- Assuming I renounce to address the question of uniformity of treatment of local definitions in this PR, I don't feel that I'll have the motivation to bring back the question of primitive projections on the table later, which makes me fear that, like other projects, it may remain pending for long.
- But maybe should I acknowledge that we are not ready to address the primitive projections issue in the short term, and that I should stop caring about it.
- But the problem is that I do care, and I'm lost in how to proceed.

Do you understand my point of view?

**Comment by ppedrot** [comment]
> Do you understand my point of view?

Even if tweaking this PR to keep backwards compatibility will make your work on primitive projections on stasis, Coq would have improved after merging this PR. So, the trade-off is pretty clear to me: either we have a minor change left to bitrot that will benefit nobody and will be probably rewritten from scratch anyways when the primitive projection mess is sorted out, or you tweak it a bit and enhance the situation for everybody. Pull requests are what their name means, *pull* *requests*. If you're not requesting anything, then we close it.

**Comment by herbelin** [comment]
> So, the trade-off is pretty clear to me: either we have a minor change left to bitrot that will benefit nobody and will be probably rewritten from scratch anyways when the primitive projection mess is sorted out, or you tweak it a bit and enhance the situation for everybody. 

I agree with that (I don't want this to bitrot and tweaking it a bit will enhance the situation for everybody). What I'm trying to get before "tweaking" this PR is a plan, or at least a schedule, in the direction of sorting out the "primitive projection mess". If I don't "use" this PR as a step, what will be the next concrete step about primitive projections? A new CEP? A call? Can I get for instance a confirmation that there is an intent to address the "mess" within a delay, say, of 6 months, with a distribution of rôles about what each of us is planning to do?

**Comment by herbelin** [comment]
@ppedrot: I followed your advice and I modified the PR so as to be compatible with the unfolding made for local definitions of primitive records.

If ok, will need a smash though.

**Comment by herbelin** [comment]
@ppedrot: I thus renounced to change the semantics of local definitions in primitive records.

I squashed and removed the useless overlay (which was anyway backwards-compatible and already committed upstream).

It is ready from my point of view.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Will merge soon.
